core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -329,6 +336,93 @@ private RexNode simplifyGenericNode(RexCall e) {
     return rexBuilder.makeCall(e.getType(), e.getOperator(), operands);
   }
 
+  /**
+   * Try to find a literal with the given value in the input list.
+   */
+  private int findLiteralIndex(List<RexNode> operands, long value) {
+    for (int i = 0; i < operands.size(); i++) {
+      if (operands.get(i).isA(SqlKind.LITERAL)) {
+        Comparable comparable = ((RexLiteral) operands.get(i)).getValue();
+        if (comparable instanceof BigDecimal && ((BigDecimal) comparable).longValue() == value) {
+          return i;
+        }
+      }
+    }
+    return -1;
+  }
+
+  private RexNode simplifyArithmetic(RexCall e) {
+    if (e.getType().getSqlTypeName().getFamily() != SqlTypeFamily.NUMERIC
+        || e.getOperands().stream()
+        .anyMatch(o -> e.getType().getSqlTypeName().getFamily() != SqlTypeFamily.NUMERIC)) {
+      // we only support simplifying numeric types
+      return simplifyGenericNode(e);
+    }
+
+    assert e.getOperands().size() == 2;
+
+    // if any operand is null, the result will be null
+    if (RexUtil.isNullLiteral(e.operands.get(0), true)
+        || RexUtil.isNullLiteral(e.operands.get(1), true)) {
+      return rexBuilder.makeNullLiteral(e.type);
+    }
```

Comment:
It might not be true. Sometimes users may expect it can throw an exception.

Reply:
Thanks for your feedback. It makes sense. This logic is removed.